O código implementa uma rede neural simples para classificar pacientes com diabetes

In [ ]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv'

#Importanto um DataSet sobre Diabete onde totas as pacientes sao mulheres

df = pd.read_csv(url)

In [ ]:
df.head()

,6,148,72,35,0,33.6,0.627,50,1
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [ ]:
df.columns

Index(['6', '148', '72', '35', '0', '33.6', '0.627', '50', '1'], dtype='object')

In [ ]:
#Mudando o nome das colunas
df.columns = [
    'Num Gestasoes',
    'Glicose',
    'Pressao Arterial',
    'Espessura da pele',
    'Insulina',
    'IMC',
    'Funcao Pedigree',
    'Idade',
    'Resultado'
]

In [ ]:
df.head()

,Num Gestasoes,Glicose,Pressao Arterial,Espessura da pele,Insulina,IMC,Funcao Pedigree,Idade,Resultado
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [ ]:
#Separando meus dados de treinamento

X = df[['Num Gestasoes', 'Glicose', 'Pressao Arterial', 'Espessura da pele', 'Insulina', 'IMC', 'Funcao Pedigree', 'Idade']]
Y = df['Resultado']

In [ ]:
from sklearn.model_selection import train_test_split

## Divide os dados em conjuntos de treinamento (80%) e teste (20%)

X_treinamento, X_teste, Y_treinamento, Y_teste = train_test_split(X, Y, test_size=0.2, random_state= 42)

In [ ]:
from sklearn.preprocessing import StandardScaler

##Normalizando os dados

scaler = StandardScaler()
X_treinamneto = scaler.fit_transform(X_treinamento)
X_teste = scaler.transform(X_teste)

In [ ]:
X.shape

(767, 8)

In [ ]:
modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,))) #Espera 8 valores de entrada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid')) # saída binária (0 ou 1)


In [ ]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [ ]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 100, batch_size=16)

Epoch 1/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.3497 - loss: 13.7631
Epoch 2/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5621 - loss: 0.9704
Epoch 3/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5917 - loss: 0.7296 
Epoch 4/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6609 - loss: 0.6576
Epoch 5/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6428 - loss: 0.6685 
Epoch 6/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6560 - loss: 0.6378
Epoch 7/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6839 - loss: 0.6318
Epoch 8/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7026 - loss: 0.5802
Epoch 9/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6769 - loss: 0.6036
Epoch 10/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6869 - loss: 0.6198
Epoch 11/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7240 - loss: 0.5756
Epoch 12/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accura

Tinha deixado apenas com uma camda oculta, mais decidi adiconar masi uma na rede e deu uma breve melhorada na acuracci

In [ ]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6583 - loss: 0.6518  


[0.6682064533233643, 0.6363636255264282]

Agora vou adicionar um DropOut, para ver se melhora ainda mais

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,)))
modelo.add(Dropout(0.3))  # desligar 30% dos neurônios dessa camada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dropout(0.3))  # outro dropout
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid'))  # camada de saída


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [ ]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 100, batch_size=16)

Epoch 1/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.5480 - loss: 11.4337
Epoch 2/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5422 - loss: 4.3900
Epoch 3/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5431 - loss: 3.2743
Epoch 4/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4867 - loss: 3.2298
Epoch 5/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5848 - loss: 1.6814
Epoch 6/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5572 - loss: 2.0392
Epoch 7/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5577 - loss: 1.1486
Epoch 8/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6180 - loss: 0.9032
Epoch 9/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6086 - loss: 1.0632
Epoch 10/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6229 - loss: 0.8549
Epoch 11/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6329 - loss: 0.8946
Epoch 12/100
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy

In [ ]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.6505 - loss: 0.6717  


[0.6943615674972534, 0.6298701167106628]

Sem camada extra -> 0.6505	0.7169
Com camada extra -> 0.6583	0.6518
Com camada + Dropout -> 0.6505	0.6717

Acredito que o DropOut funcionou evitanto o Overfitting, por isso a acuraccia abaixou

# Otimizando o modelo

Diminuindo um pouco o DropOut
E aumentando o numero de epocas de treinamento

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

modelo = Sequential()
modelo.add(Dense(32, activation='relu', input_shape=(8,)))
modelo.add(Dropout(0.2))  # desligar 30% dos neurônios dessa camada
modelo.add(Dense(16, activation='relu'))
modelo.add(Dropout(0.2))  # outro dropout
modelo.add(Dense(8, activation='relu'))
modelo.add(Dense(1, activation='sigmoid'))  # camada de saída


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
modelo.compile(
    optimizer='adam',
    loss='binary_crossentropy',    # Função para classificação binária
    metrics=['accuracy']
)

In [ ]:
modelo.fit(X_treinamento, Y_treinamento, epochs = 200, batch_size=16)

Epoch 1/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6552 - loss: 8.2704
Epoch 2/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5745 - loss: 1.9972
Epoch 3/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5520 - loss: 1.2443
Epoch 4/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5814 - loss: 1.1590
Epoch 5/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5506 - loss: 1.0352
Epoch 6/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6011 - loss: 0.9296
Epoch 7/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5888 - loss: 0.8168
Epoch 8/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5619 - loss: 0.7745
Epoch 9/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5934 - loss: 0.7290
Epoch 10/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6401 - loss: 0.8053 
Epoch 11/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6534 - loss: 0.7107
Epoch 12/200
39/39 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy

In [ ]:
modelo.evaluate(X_teste, Y_teste) #avaliando desempenho do modelo

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6505 - loss: 0.6889 


[0.7138802409172058, 0.6298701167106628]

Perda = 0.713

Acuracia = 0.62